### PR TITLE
Async scheduled tasks

### PR DESCRIPTION
**Pull request by pxp9**
Solve #40

### CODE REVIEW

File: src/asynk/async_queue.rs [hunk 1]
```diff
@@ -443,6 +576,50 @@ where
...
+    ) -> Result<PeriodicTask, AsyncQueueError> {
+        let mut connection = self.pool.get().await?;
+        let mut transaction = connection.transaction().await?;
+
+        let periodic_task = Self::schedule_next_task_query(&mut transaction, periodic_task).await?;
+
+        transaction.commit().await?;
+
+        Ok(periodic_task)
+    }
```
Comment (ayrat555) [inline src/asynk/async_queue.rs:607]: 🎨  Let's use one line between methods

**Review by ayrat555 — APPROVED** [review]

**Review by ayrat555 — APPROVED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -45,6 +45,12 @@ version = "0.8"
 features = ["with-serde_json-1" , "with-uuid-0_8" , "with-chrono-0_4"]
 optional = true
 
+[dependencies.futures]
```
[the remark this answers is not in the document]
Comment (pxp9) [reply Cargo.toml]: You are right

File: src/asynk/async_scheduler.rs [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    async fn process_task(&mut self, task: PeriodicTask) -> Result<(), Error> {
+        match task.scheduled_at {
+            None => {
+                self.queue.schedule_next_task(task).await?;
+            }
+            Some(_) => {
+                let metadata = task.metadata.clone();
+                // known issue,  hardcode common to try if rest of code works
```
Comment (pxp9) [inline src/asynk/async_scheduler.rs:77]: Have to  talk about how to access task type, maybe one scheduler is will be for each task type, or somehow keep task type in PeriodicTask schema in database

File: src/asynk/async_scheduler.rs [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        pub number: u16,
+    }
+
+    #[typetag::serde]
+    #[async_trait(?Send)]
+    impl AsyncRunnable for AsyncScheduledTask {
+        async fn run(&self, _queueable: &mut dyn AsyncQueueable) -> Result<(), Error> {
+            Ok(())
+        }
+        /*fn task_type(&self) -> String {
```
Comment (pxp9) [inline src/asynk/async_scheduler.rs:139]: we will delete comment here when task type is fixed.

File: src/asynk/async_scheduler.rs [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+        let check_period: u64 = 1;
+        let error_margin_seconds: u64 = 2;
+
+        let mut scheduler = Scheduler::builder()
+            .check_period(check_period)
+            .error_margin_seconds(error_margin_seconds)
+            .queue(&mut test as &mut dyn AsyncQueueable)
+            .build();
+        // Scheduler start tricky not loop :)
```
Comment (pxp9) [inline src/asynk/async_scheduler.rs:167]: this comment is optional , if you want, i will delete

**Review by pxp9 — COMMENTED** [review]
Thinking about task type in scheduler.

File: src/asynk/async_queue.rs [hunk 6]
```diff
@@ -139,43 +134,75 @@ pub trait AsyncQueueable {
...
 where
     Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
     <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
     <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
     <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
 {
+    #[builder(setter(into))]
     pool: Pool<PostgresConnectionManager<Tls>>,
+    #[builder(setter(into))]
+    duplicated_tasks: bool,
```
Comment (ayrat555) [inline src/asynk/async_queue.rs:166]: maybe a default value can be added for this field?

File: src/asynk/async_scheduler.rs (hunk 3 above)
Comment (ayrat555) [reply src/asynk/async_scheduler.rs:77]: let's restart a task from json metadata

File: src/asynk/async_queue.rs [hunk 7]
```diff
@@ -613,7 +847,10 @@ mod async_queue_tests {
         let mut connection = pool.get().await.unwrap();
         let transaction = connection.transaction().await.unwrap();
 
-        let mut test = AsyncQueueTest { transaction };
+        let mut test = AsyncQueueTest {
```
Comment (ayrat555) [inline src/asynk/async_queue.rs:850]: let's use a builder to create this struct

File: src/asynk/async_scheduler.rs [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    #[builder(setter(into))]
+    pub queue: &'a mut dyn AsyncQueueable,
+    #[builder(default = 0, setter(into))]
+    pub number_of_restarts: u32,
+}
+
+impl<'a> Scheduler<'a> {
+    #[async_recursion(?Send)]
+    pub async fn start(&mut self) -> Result<(), Error> {
+        let task_res = self.schedule_loop().await;
```
Comment (ayrat555) [inline src/asynk/async_scheduler.rs:25]: let's add 1 second of sleep after this step finishes.

so a system (maybe db) has time to recover

**Review by ayrat555 — COMMENTED** [review]
Good work

File: src/asynk/async_scheduler.rs (hunk 3 above)
Comment (pxp9) [reply src/asynk/async_scheduler.rs]: Solve in [a689d02](https://github.com/ayrat555/fang/pull/47/commits/a689d022634d4fd8ba3e4690dc1557b24030bc00)

File: src/asynk/async_scheduler.rs (hunk 4 above)
Comment (pxp9) [reply src/asynk/async_scheduler.rs]: Solve in [a689d02](https://github.com/ayrat555/fang/pull/47/commits/a689d022634d4fd8ba3e4690dc1557b24030bc00)

File: src/asynk/async_scheduler.rs (hunk 5 above)
Comment (pxp9) [reply src/asynk/async_scheduler.rs:178]: Solve in [a689d02](https://github.com/ayrat555/fang/pull/47/commits/a689d022634d4fd8ba3e4690dc1557b24030bc00)

File: src/asynk/async_queue.rs (hunk 6 above)
Comment (pxp9) [reply src/asynk/async_queue.rs:166]: default true ???

Comment (pxp9) [reply src/asynk/async_queue.rs:166]: False default

File: src/asynk/async_scheduler.rs (hunk 8 above)
Comment (pxp9) [reply src/asynk/async_scheduler.rs:26]: Solve in [e1d31d2](https://github.com/ayrat555/fang/pull/47/commits/e1d31d215656077b385a07f4582a30a6d595a3c5)

File: src/asynk/async_queue.rs (hunk 7 above)
Comment (pxp9) [reply src/asynk/async_queue.rs]: Solve in [e1d31d2](https://github.com/ayrat555/fang/pull/47/commits/e1d31d215656077b385a07f4582a30a6d595a3c5)

File: src/asynk/async_queue.rs (hunk 6 above)
Comment (pxp9) [reply src/asynk/async_queue.rs:166]: Solve in [e1d31d2](https://github.com/ayrat555/fang/pull/47/commits/e1d31d215656077b385a07f4582a30a6d595a3c5)

**Review by ayrat555 — APPROVED** [review]

**Pull request merged**